File: app/src/main/java/com/seemingamusing/android/playground/footer/FooterBarBehavior.java
package com.seemingamusing.android.playground.footer;

import android.content.Context;
import android.support.design.widget.AppBarLayout;
import android.support.design.widget.CoordinatorLayout;
import android.util.AttributeSet;
import android.util.Log;
import android.view.View;
import com.seemingamusing.android.playground.R;

public class FooterBarBehavior extends AppBarLayout.ScrollingViewBehavior {

  public FooterBarBehavior() {
  }

  public FooterBarBehavior(Context context, AttributeSet attrs) {
    super(context, attrs);
  }

  @Override public boolean layoutDependsOn(CoordinatorLayout parent, View child, View dependency) {
    boolean isFooterBar = isFooterBar(dependency);
    if (isFooterBar) {
      adjustChildBottomMargin(child, dependency);
    }
    return isFooterBar || super.layoutDependsOn(parent, child, dependency);
  }

  private void adjustChildBottomMargin(View child, View dependency) {
    CoordinatorLayout.LayoutParams lp = (CoordinatorLayout.LayoutParams) child.getLayoutParams();
    int height = 0;
    if (dependency.getVisibility() == View.VISIBLE) {
      height = dependency.getHeight();
    }
    lp.bottomMargin = height;
    Log.d(getClass().getSimpleName(), "Height: " + height);
  }

  private boolean isFooterBar(View dependency) {
    return dependency.getId() == R.id.footer_bar;
  }

  @Override
  public boolean onDependentViewChanged(CoordinatorLayout parent, View child, View dependency) {
    if (isFooterBar(dependency)) {
      adjustChildBottomMargin(child, dependency);
    }
    return super.onDependentViewChanged(parent, child, dependency);
  }
}

File: app/src/main/java/com/seemingamusing/android/playground/footer/FooterBarBehavior.kt
package com.seemingamusing.android.playground.footer

import android.content.Context
import android.support.design.widget.AppBarLayout
import android.support.design.widget.CoordinatorLayout
import android.util.AttributeSet
import android.util.Log
import android.view.View
import com.seemingamusing.android.playground.R

class FooterBarBehavior : AppBarLayout.ScrollingViewBehavior {

  constructor()
  constructor(context: Context, attrs: AttributeSet) : super(context, attrs)

  override fun layoutDependsOn(parent: CoordinatorLayout, child: View, dependency: View) =
      isFooterBar(dependency).also { isFooterBar ->
        if (isFooterBar) adjustChildBottomMargin(child, dependency)
      } || super.layoutDependsOn(parent, child, dependency)

  private fun isFooterBar(dependency: View) = dependency.id == R.id.footer_bar

  private fun adjustChildBottomMargin(child: View, dependency: View) {
    val height = when (dependency.visibility) {
      View.VISIBLE -> dependency.height
      else -> 0
    }
    (child.layoutParams as CoordinatorLayout.LayoutParams).bottomMargin = height
    Log.d(javaClass.simpleName, "Height: $height")
  }

  override fun onDependentViewChanged(parent: CoordinatorLayout?, child: View, dependency: View): Boolean {
    if (isFooterBar(dependency)) adjustChildBottomMargin(child, dependency)
    return super.onDependentViewChanged(parent, child, dependency)
  }
}